opsdroid/database/matrix/__init__.py

```diff
@@ -0,0 +1,269 @@
+    "should_encrypt": bool,
+}
+
+
+class DatabaseMatrix(Database):
+    """A module for opsdroid to allow memory to persist in matrix room state."""
+
+    def __init__(self, config, opsdroid=None):
+        """Start the database connection."""
+        super().__init__(config, opsdroid=opsdroid)
+        self.name = "matrix"
+        self.room = config.get("default_room", "main")
+        self._single_state_key = config.get("single_state_key", True)
+        self.should_encrypt = config.get("should_encrypt", True)
+        self._event_type = "dev.opsdroid.database"
+        self.should_migrate = True
+
+        _LOGGER.debug("Loaded matrix database connector.")
+
+    @property
+    def connector(self):
+        """Define the initiated matrix connector."""
+        return self.opsdroid.get_connector("matrix")
+
+    @property
+    def room_id(self):
+        """Define ID of the room to store data in."""
+        return self.connector.lookup_target(self.room)
+
+    @property
+    def is_room_encrypted(self):
+        """Check if room is encrypted."""
+        return (
+            self.connector._allow_encryption
+            and self.room_id in self.connector.connection.store.load_encrypted_rooms()
+        )
+
+    async def migrate_database(self):
+        """Migrate existing 'opsdroid.database' state event to 'dev.opsdroid.database' event."""
+
+        data = await self.connector.connection.room_get_state(room_id=self.room_id)
+        if isinstance(data, RoomGetStateError):
+            _LOGGER.error(
+                f"Error migrating from opsdroid.database to {self._event_type} in room {self.room_id}: {data.message}({data.status_code})"
+            )
+            return
+
+        if not any(event["type"] == self._event_type for event in data.events):
+            for event in data.events:
+                if event["type"] == "opsdroid.database" and event["content"]:
+                    if not self._single_state_key:
+                        event["content"] = {event["state_key"]: event["content"]}
+                    await self.opsdroid.send(
+                        MatrixStateEvent(
+                            self._event_type,
+                            content=event["content"],
+                            target=self.room_id,
+                            connector=self.connector,
+                            state_key=event["state_key"],
+                        )
+                    )
+                    await self.connector.connection.room_redact(
+                        room_id=self.room_id, event_id=event["event_id"],
+                    )
+
+        self.should_migrate = False
+
+    async def connect(self):
+        """Connect to the database."""
+
+        _LOGGER.info("Matrix Database connector initialised.")
+
+    async def put(self, key, value):
+        """Insert or replace a value into the database for a given key."""
+
+        if self.should_migrate:
+            await self.migrate_database()
```

Comment:
Is it worth just doing this on startup rather than checking if it's been done every time `get()` or `put()` is called?

Reply:
you could be in a new room everytime with the decorator.